File: Data Structures & Algorithms/trapping-rain-water/submission-1.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        left = 0
        right = n - 1
        prefix = height[left]
        postfix = height[right]
        ans = 0
        while left <= right:
            if prefix < postfix:
                ans += prefix - height[left]
                left += 1
                prefix = max(prefix, height[left])
            else:
                ans += postfix - height[right]
                right -= 1
                postfix = max(postfix, height[right])
        return ans

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    TreeNode ans;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        lca(root, p, q);
        return ans;
    }

    private boolean lca(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null) return false;
        boolean left = lca(root.left, p, q);
        boolean right = lca(root.right, p, q);
        if ((left == true && right == true) || ((root.val == p.val || root.val == q.val) && (left || right))) {
            ans = root;
        }
        return (left || right || root.val == p.val || root.val == q.val);
    }
}
